portfolio/src/main/webapp/js/background.js

```diff
@@ -0,0 +1,375 @@
+const HEIGHT_TO_WIDTH = 0.3;
+/**
+ * Runs initial set up of the background and background canvas.
+ */
+function setup() {
+  const FRAMERATE = 60;
+  const backgroundContainer = document.getElementById("background-container");
+
+  const backgroundCanvas =
+    createCanvas(backgroundContainer.clientWidth,
+      backgroundContainer.clientWidth * HEIGHT_TO_WIDTH);
+  backgroundCanvas.parent("background-container");
+
+  frameRate(FRAMERATE);
+
+  initTank();
+}
+
+let ballTank;
+/**
+ * Initializes a tank for the background.
+ */
+function initTank() {
+  const EFFECT_RAD_TO_WIDTH = 0.1;
+  const BALL_SIZE_TO_WIDTH = 0.01;
+  const NUM_BALLS = 25;
+  const INIT_BALL_VEL = 1;
+
+  const ballSize = BALL_SIZE_TO_WIDTH * width;
+  ballTank = new Tank(width, height, EFFECT_RAD_TO_WIDTH * width, ballSize);
+
+  for (let i = 0; i < NUM_BALLS; i++) {
+    const randomPos = createVector(Math.random() * ballTank.width,
+      Math.random() * ballTank.height);
+
+    const randomVel = p5.Vector.random2D();
+    randomVel.mult(INIT_BALL_VEL);
+
+    const newBall = new Ball(randomPos, randomVel);
+    ballTank.addBall(newBall);
+  }
+}
+
+/**
+ * Resizes the canvas when the window is resized.
+ */
+function windowResized() {
+  const canvasContainer = document.getElementById("background-container");
+  resizeCanvas(canvasContainer.clientWidth,
+    canvasContainer.clientWidth * HEIGHT_TO_WIDTH);
+}
+
+/**
+ * Called every draw loop; used for updating game logic
+ * and drawing to the screen.
+ */
+function draw() {
+  background(255);
+  ballTank.update();
+  ballTank.draw();
+}
+
+/**
+ * Tank that holds, updates, and draws statically-charged balls.
+ */
+class Tank {
+  /**
+   * @param {number} width
+   * @param {number} height
+   * @param {number} staticRadius
+   * @param {number} ballSize
+   */
+  constructor(width, height, staticRadius, ballSize) {
+    /**
+     * Height of the tank.
+     * @type {number} @const
+     */
+    this.height = height;
+
+    /**
+     * Width of the tank.
+     * @type {number} @const
+     */
+    this.width = width;
+
+    /**
+     * Effective radius of static effect.
+     * @type {number} @const
+     */
+    this.staticRadius = staticRadius;
+
+    /**
+     * Effective diameter of static effect.
+     * @type {number} @const
+     */
+    this.staticDiameter = staticRadius * 2;
+
+    /**
+     * The size of the static balls.
+     * @type {number} @const
+     */
+    this.ballSize = ballSize;
+
+    /**
+     * Spatial hashing grid used to reduce calls to balls
+     * beyond effective radius.
+     * @type {Object.<string, Array<Ball>>} @private
+     */
+    this._spatialHash = {};
```

Comment:
I have two comments:
 1. Is this optimization necessary? If everything runs smoothly without it (and will continue to run smoothly with the maximum number of balls you'd expect to ever want), you should remove it so that your code is easier to read.
 2. If this optimization is necessary, you should create a separate object that handles the logic for inserting/deleting/looking up items in the spatialHash. Separating out _updateHash, _placeInHash, _removeFromHash, and _getSpatialHashPos into a new object would make it clear what is hashing logic and what is not.

Reply:
You may also want to move getNearby into that hash object also.

Reply:
Yeah you're right that this optimization is not really necessary. Guess I got carried away haha. Removing it will significantly improve readability.